### PR TITLE
feat(sdk): deep equality

### LINKED ISSUE #2938: Json deep equality not implemented
### I tried this:

```js
assert((Json { "foo": 1, "bar": 2 }) == Json { "foo": 1, "bar": 2 });
```

### This happened:

```
ERROR: assertion failed: '(Object.freeze({"foo":1,"bar":2}) === Object.freeze({"foo":1,"bar":2}))'

target/equality.wsim.125546.tmp/.wing/preflight.js:8
     constructor(scope, id) {
       super(scope, id);
>>     {((cond) => {if (!cond) throw new Error(`assertion failed: '(Object.freeze({"foo":1,"bar":2}) === Object.freeze({"foo":1,"bar":2}))'`)})((Object.freeze({"foo":1,"bar":2}) === Object.freeze({"foo":1,"bar":2})))};
     }
   }

    at compile (/Users/chrisr/.volta/tools/image/packages/winglang/lib/node_modules/winglang/dist/commands/compile.js:126:19)
```

### I expected this:

It compiles

### Is there a workaround?

Stringify your Json and then compare it:

```js
assert(Json.stringify(Json { "foo": 1, "bar": 2 }) == Json.stringify(Json { "foo": 1, "bar": 2 }));
```

### Component

Compiler

### Wing Version

0.19.14

### Wing Console Version

_No response_

### Node.js Version

18.14.1

### Platform(s)

MacOS

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #2939: Struct deep equality not implemented
### I tried this:

```js
struct Cat {
  name: str;
  age: num;
}

let cat1 = Cat { name: "Mittens", age: 3 };
let cat2 = Cat { name: "Mittens", age: 3 };

assert(cat1 == cat2); // fields and types match
```

### This happened:

```
ERROR: assertion failed: '(cat1 === cat2)'

target/equality.wsim.725637.tmp/.wing/preflight.js:24
       "age": 3,}
       ;
>>     {((cond) => {if (!cond) throw new Error(`assertion failed: '(cat1 === cat2)'`)})((cat1 === cat2))};
       {((cond) => {if (!cond) throw new Error(`assertion failed: '(cat1 !== cat3)'`)})((cat1 !== cat3))};
     }
```

### I expected this:

It compiles

### Is there a workaround?

Write a custom equality function that checks if each field is equal

### Component

Compiler

### Wing Version

0.19.14

### Wing Console Version

_No response_

### Node.js Version

18.14.1

### Platform(s)

MacOS

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #2940: Collection type deep equality not implemented
### I tried this:

```js
assert(Array<num>[1, 2, 3] == Array<num>[1, 2, 3]);
assert(Array<num>[1, 2, 3] != Array<num>[3, 2, 1]);

assert(Map<str>{"a": "1", "b": "2"} == Map<str>{"a": "1", "b": "2"});
assert(Map<str>{"a": "1", "b": "2"} == Map<str>{"b": "2", "a": "1"});

assert(Set<num>{1, 2, 3} == Set<num>{1, 2, 3});
assert(Set<num>{1, 2, 3} == Set<num>{3, 2, 1});
```

### This happened:

```
ERROR: assertion failed: '(Object.freeze([1, 2, 3]) === Object.freeze([1, 2, 3]))'

target/equality.wsim.166538.tmp/.wing/preflight.js:8
     constructor(scope, id) {
       super(scope, id);
>>     {((cond) => {if (!cond) throw new Error(`assertion failed: '(Object.freeze([1, 2, 3]) === Object.freeze([1, 2, 3]))'`)})((Object.freeze([1, 2, 3]) === Object.freeze([1, 2, 3])))};
       {((cond) => {if (!cond) throw new Error(`assertion failed: '(Object.freeze([1, 2, 3]) !== Object.freeze([3, 2, 1]))'`)})((Object.freeze([1, 2, 3]) !== Object.freeze([3, 2, 1])))};
       {((cond) => {if (!cond) throw new Error(`assertion failed: '(Object.freeze({"a":"1","b":"2"}) === Object.freeze({"a":"1","b":"2"}))'`)})((Object.freeze({"a":"1","b":"2"}) === Object.freeze({"a":"1","b":"2"})))};
```

### I expected this:

It compiles

### Is there a workaround?

Write a custom equality function that checks the array lengths and their elements

### Component

Compiler

### Wing Version

0.19.14

### Wing Console Version

_No response_

### Node.js Version

18.14.1

### Platform(s)

MacOS

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by marciocadev**
Closes #2938
Closes #2939
Closes #2940

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [ ] Description explains motivation and solution
- [x] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

**Comment by Chriscbr** [comment]
Hey @marciocadev! 👋 Thanks a lot for diving into this issue and taking the initiative to work on it. I really appreciate your effort!

I've gone through the code changes, and it's looking good overall. However, there's a crucial aspect we need to address before we can merge this pull request. As part of the compiler, we need to ensure that deep equality is automatically performed whenever two Json values are checked for equality using `==` in Wing.

Currently, the implementation seems to rely on the user explicitly calling a special method like "isEquals" to achieve deep equality. Instead, we should strive for a seamless experience for our users. They shouldn't have to remember to use an extra method for something as fundamental as checking equality.

To achieve this, I suggest looking into modifying how we emit JavaScript for the "==" operator: https://github.com/winglang/wing/blob/11b57b97a366d793b53fee9f3b5bde5684bd2435/libs/wingc/src/jsify.rs#L469C6-L469C6. I think we might be able to use the built-in `assert.strictDeepEquals` method from Node.js to get deep equality checking to work. It should serve our purpose perfectly and eliminate the need for users to call a specific method for deep equality checks.

Additionally, let's ensure we've added several examples of equality checking in the `examples/` directory that include cases where `Json` values are being compared to each other and `MutJson` values are being compared to each other.

If you have any questions or need further guidance, feel free to reach out. I'm looking forward to seeing this enhancement in action!

**Comment by hasanaburayyan** [comment]
@marciocadev🚀 Im super excited to see this issue getting some traction. I agree with @Chriscbr that we should do some work in the jsification to give that easy to use experience, limiting the cognitive friction that poor Json type already has 😄

**Comment by marciocadev** [comment]
@hasanaburayyan and @Chriscbr 

I made a somewhat radical change in the comparison, now it's truly checking the value regardless of the variable type, but I found the code became too verbose.

**Comment by marciocadev** [comment]
hi @hasanaburayyan, there's something I don't understand.

When I compare a Json object with an object that has the same value in a MutJson, there's no error indicating they are different types.

However, when I compare a Set with a MutSet, this error is displayed.

Why does the Json object have a different error handling approach?
image
image

**Comment by Chriscbr** [comment]
@marciocadev Indeed that's a bug - I found the relevant issue and added a comment here: https://github.com/winglang/wing/issues/2867. It should be okay to compare Json and MutJson values to the best of my knowledge 🙂

File: examples/tests/valid/deep_equality.w [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  assert(jsonA == jsonA);
+  assert(jsonA == jsonB);
+}
+
+test "Json with different values" {
+  try {
+    assert(jsonA == jsonC);
+  } catch e {
+    assert(e == "assertion failed: jsonA == jsonC");
+  }
```
Comment (Chriscbr) [inline examples/tests/valid/deep_equality.w:51]: Let's fix this in all of the places
```suggestion
  assert(jsonA != jsonC);
```

File: examples/tests/valid/deep_equality.w [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+//-----------------------------------------------------------------------------
+// Set
+//-----------------------------------------------------------------------------
+let setA = { 1,2,3 };
+let setB = MutSet<num>{ 1,2,3 };
+let setC = { 4,5,6 };
+
+test "Set types with the same value" {
+  assert(setA == setA);
+  assert(setA == setB.copy());
```
Comment (Chriscbr) [inline examples/tests/valid/deep_equality.w:63]:
```suggestion
  // TODO https://github.com/winglang/wing/issues/2867
  assert(setA == setB.copy());
```

**Review by Chriscbr — APPROVED** [review]
Awesome! Just a small comment on the test, otherwise it looks ready to ship :)

**Pull request merged**